File: api/src/main/java/dev/vibris/api/ReloadResult.java
package dev.vibris.api;

import java.util.List;
import java.util.Objects;

public record ReloadResult(boolean successful, List<Diagnostic> diagnostics) {
    public ReloadResult {
        diagnostics = List.copyOf(diagnostics);
    }

    public static ReloadResult success(List<Diagnostic> diagnostics) {
        return new ReloadResult(true, diagnostics);
    }

    public static ReloadResult failure(List<Diagnostic> diagnostics) {
        return new ReloadResult(false, diagnostics);
    }

    public record Diagnostic(Severity severity, String source, int line, String message) {
        public Diagnostic {
            severity = Objects.requireNonNull(severity, "severity");
            source = Objects.requireNonNull(source, "source");
            message = Objects.requireNonNull(message, "message");
            if (line < 0) throw new IllegalArgumentException("line must not be negative");
        }
    }

    public enum Severity {
        INFO,
        WARNING,
        ERROR
    }
}

File: api/src/main/kotlin/dev/vibris/api/ReloadResult.kt
package dev.vibris.api

@JvmRecord
data class ReloadResult(
    val successful: Boolean,
    val activeStatePreserved: Boolean,
    val effectiveSettings: EffectiveShaderSettings,
    @field:DefensiveSnapshot val diagnostics: List<Diagnostic>,
) {
    init {
        require(!successful || !activeStatePreserved) {
            "A successful reload establishes a new verified active state"
        }
    }

    @JvmRecord
    data class Diagnostic(
        val severity: Severity,
        val source: String,
        val line: Int,
        val message: String,
    ) {
        init {
            require(line >= 0) { "line must not be negative" }
        }
    }

    enum class Severity {
        INFO,
        WARNING,
        ERROR,
    }

    companion object {
        @JvmStatic
        fun success(
            effectiveSettings: EffectiveShaderSettings,
            diagnostics: List<Diagnostic>,
        ): ReloadResult = ReloadResult(true, false, effectiveSettings, diagnostics)

        @JvmStatic
        fun failure(diagnostics: List<Diagnostic>): ReloadResult =
            ReloadResult(false, false, EffectiveShaderSettings.empty(), diagnostics)

        @JvmStatic
        fun failurePreservingActiveState(
            effectiveSettings: EffectiveShaderSettings,
            diagnostics: List<Diagnostic>,
        ): ReloadResult = ReloadResult(false, true, effectiveSettings, diagnostics)
    }
}